blockchain/v2/reactor.go

```diff
@@ -1,118 +1,521 @@
+	cdc.RegisterInterface((*BlockchainMessage)(nil), nil)
+	cdc.RegisterConcrete(&bcBlockRequestMessage{}, "tendermint/blockchain/BlockRequest", nil)
+	cdc.RegisterConcrete(&bcBlockResponseMessage{}, "tendermint/blockchain/BlockResponse", nil)
+	cdc.RegisterConcrete(&bcNoBlockResponseMessage{}, "tendermint/blockchain/NoBlockResponse", nil)
+	cdc.RegisterConcrete(&bcStatusResponseMessage{}, "tendermint/blockchain/StatusResponse", nil)
+	cdc.RegisterConcrete(&bcStatusRequestMessage{}, "tendermint/blockchain/StatusRequest", nil)
 }
 
-func (r *Reactor) Receive(event Event) {
-	// XXX: decode and serialize write events
-	// TODO: backpressure
+func decodeMsg(bz []byte) (msg BlockchainMessage, err error) {
+	if len(bz) > maxMsgSize {
+		return msg, fmt.Errorf("msg exceeds max size (%d > %d)", len(bz), maxMsgSize)
+	}
+	err = cdc.UnmarshalBinaryBare(bz, &msg)
+	return
+}
+
+// Receive implements Reactor by handling different message types.
+func (r *BlockchainReactor) Receive(chID byte, src p2p.Peer, msgBytes []byte) {
+	msg, err := decodeMsg(msgBytes)
+	if err != nil {
+		r.logger.Error("error decoding message",
+			"src", src.ID(), "chId", chID, "msg", msg, "err", err, "bytes", msgBytes)
+		_ = r.reporter.Report(behaviour.BadMessage(src.ID(), err.Error()))
+		return
+	}
+
+	if err = msg.ValidateBasic(); err != nil {
+		r.logger.Error("peer sent us invalid msg", "peer", src, "msg", msg, "err", err)
+		_ = r.reporter.Report(behaviour.BadMessage(src.ID(), err.Error()))
+		return
+	}
+
+	r.logger.Debug("Receive", "src", src.ID(), "chID", chID, "msg", msg)
+
+	switch msg := msg.(type) {
+	case *bcStatusRequestMessage:
+		if err := r.io.sendStatusResponse(r.store.Height(), src.ID()); err != nil {
+			r.logger.Error("Could not send status message to peer", "src", src)
+		}
+
+	case *bcBlockRequestMessage:
+		block := r.store.LoadBlock(msg.Height)
+		if block != nil {
+			if err = r.io.sendBlockToPeer(block, src.ID()); err != nil {
+				r.logger.Error("Could not send block message to peer: ", err)
+			}
+		} else {
+			r.logger.Info("peer asking for a block we don't have", "src", src, "height", msg.Height)
+			peerID := src.ID()
+			if err = r.io.sendBlockNotFound(msg.Height, peerID); err != nil {
+				r.logger.Error("Couldn't send block not found: ", err)
+			}
+		}
+
+	case *bcStatusResponseMessage:
+		r.events <- bcStatusResponse{peerID: src.ID(), height: msg.Height}
+
+	case *bcBlockResponseMessage:
+		r.events <- bcBlockResponse{
+			peerID: src.ID(),
+			block:  msg.Block,
+			size:   int64(len(msgBytes)),
+			time:   time.Now(),
+		}
+
+	case *bcNoBlockResponseMessage:
+		r.events <- bcNoBlockResponse{peerID: src.ID(), height: msg.Height, time: time.Now()}
+	}
+}
+
+// AddPeer implements Reactor interface
+func (r *BlockchainReactor) AddPeer(peer p2p.Peer) {
+	err := r.io.sendStatusResponse(r.store.Height(), peer.ID())
+	if err != nil {
+		r.logger.Error("Could not send status message to peer new", "src", peer.ID, "height", r.SyncHeight())
```

Comment:
is it ok to continue if we can't send something to a peer? just want to double check

Reply:
I think it's ok. The peer will send a `statusRequest` later.